File: lib/redis/session.ts
/**
 * Redis Session Manager
 *
 * Session management for WebSocket synchronization using Redis.
 * Stores session state, client connections, and provides pub/sub for real-time updates.
 *
 * @module lib/redis/session
 */

import type { Song, DisplaySettings } from "@/lib/websocket/client";
import { redisGet, redisSet, redisDel, redisSAdd, redisSMembers } from "./client";

// ============================================================================
// Types
// ============================================================================

export interface SessionState {
  sessionId: string;
  currentSong: Song | null;
  currentLineIndex: number;
  isPlaying: boolean;
  settings: DisplaySettings;
  controllerCount: number;
  displayCount: number;
  createdAt: number;
  updatedAt: number;
}

export interface ClientInfo {
  clientId: string;
  role: "controller" | "display" | "admin";
  userId: string | null;
  joinedAt: number;
}

// ============================================================================
// Constants
// ============================================================================

const SESSION_PREFIX = "session:";
const SESSION_CLIENTS_PREFIX = "session:clients:";
const SESSION_TTL = 3600; // 1 hour in seconds

// ============================================================================
// Session CRUD Operations
// ============================================================================

/**
 * Get a session by ID
 */
export async function getSession(sessionId: string): Promise<SessionState | null> {
  return await redisGet<SessionState>(`${SESSION_PREFIX}${sessionId}`);
}

/**
 * Create a new session
 */
export async function createSession(
  sessionId: string,
  initialState: Partial<Omit<SessionState, "sessionId" | "createdAt" | "updatedAt">> = {}
): Promise<SessionState> {
  const now = Date.now();

  const session: SessionState = {
    sessionId,
    currentSong: null,
    currentLineIndex: 0,
    isPlaying: false,
    settings: {
      displayLines: 4,
      fontSize: 24,
      fontFamily: "Inter",
      theme: "dark",
      showBackground: true,
      backgroundColor: "#000000",
      textColor: "#ffffff",
      highlightColor: "#0ea5e9",
      autoScroll: true,
      scrollDuration: 300,
      enableAnimation: true,
    },
    controllerCount: 0,
    displayCount: 0,
    createdAt: now,
    updatedAt: now,
    ...initialState,
  };

  await redisSet(`${SESSION_PREFIX}${sessionId}`, session, SESSION_TTL);

  return session;
}

/**
 * Update a session
 */
export async function updateSession(
  sessionId: string,
  updates: Partial<Omit<SessionState, "sessionId" | "createdAt" | "updatedAt">>
): Promise<SessionState | null> {
  const current = await getSession(sessionId);

  if (!current) {
    return null;
  }

  const updated: SessionState = {
    ...current,
    ...updates,
    updatedAt: Date.now(),
  };

  await redisSet(`${SESSION_PREFIX}${sessionId}`, updated, SESSION_TTL);

  return updated;
}

/**
 * Delete a session
 */
export async function deleteSession(sessionId: string): Promise<boolean> {
  // Delete session state
  await redisDel(`${SESSION_PREFIX}${sessionId}`);

  // Delete clients set
  await redisDel(`${SESSION_CLIENTS_PREFIX}${sessionId}`);

  return true;
}

// ============================================================================
// Session Client Management
// ============================================================================

/**
 * Add a client to a session
 */
export async function addClientToSession(
  sessionId: string,
  clientInfo: ClientInfo
): Promise<void> {
  const clientsKey = `${SESSION_CLIENTS_PREFIX}${sessionId}`;

  // Add to clients set (store as JSON)
  await redisSAdd(clientsKey, JSON.stringify(clientInfo));

  // Update session counts
  const session = await getSession(sessionId);
  if (session) {
    const controllerCount = await getClientCountByRole(sessionId, "controller");
    const displayCount = await getClientCountByRole(sessionId, "display");

    await updateSession(sessionId, {
      controllerCount,
      displayCount,
    });
  }
}

/**
 * Remove a client from a session
 */
export async function removeClientFromSession(
  sessionId: string,
  clientId: string
): Promise<void> {
  const clientsKey = `${SESSION_CLIENTS_PREFIX}${sessionId}`;

  // Get all clients and filter out the one to remove
  const clients = await getSessionClients(sessionId);
  const updatedClients = clients.filter((c) => c.clientId !== clientId);

  // Delete and re-add all clients
  await redisDel(clientsKey);
  for (const client of updatedClients) {
    await redisSAdd(clientsKey, JSON.stringify(client));
  }

  // Update session counts
  const session = await getSession(sessionId);
  if (session) {
    const controllerCount = await getClientCountByRole(sessionId, "controller");
    const displayCount = await getClientCountByRole(sessionId, "display");

    await updateSession(sessionId, {
      controllerCount,
      displayCount,
    });
  }

  // Clean up session if no clients left
  if (updatedClients.length === 0) {
    await deleteSession(sessionId);
  }
}

/**
 * Get all clients in a session
 */
export async function getSessionClients(sessionId: string): Promise<ClientInfo[]> {
  const clientsKey = `${SESSION_CLIENTS_PREFIX}${sessionId}`;
  const members = await redisSMembers(clientsKey);

  return members.map((member) => JSON.parse(member) as ClientInfo);
}

/**
 * Get client count by role in a session
 */
async function getClientCountByRole(
  sessionId: string,
  role: "controller" | "display" | "admin"
): Promise<number> {
  const clients = await getSessionClients(sessionId);
  return clients.filter((c) => c.role === role).length;
}

/**
 * Check if a client is in a session
 */
export async function isClientInSession(
  sessionId: string,
  clientId: string
): Promise<boolean> {
  const clientsKey = `${SESSION_CLIENTS_PREFIX}${sessionId}`;
  const members = await redisSMembers(clientsKey);

  return members.some((member) => {
    const client = JSON.parse(member) as ClientInfo;
    return client.clientId === clientId;
  });
}

// ============================================================================
// Session State Updates
// ============================================================================

/**
 * Update current song in session
 */
export async function updateSessionSong(
  sessionId: string,
  song: Song | null
): Promise<SessionState | null> {
  return await updateSession(sessionId, {
    currentSong: song,
    currentLineIndex: 0,
  });
}

/**
 * Update current line index in session
 */
export async function updateSessionLine(
  sessionId: string,
  lineIndex: number
): Promise<SessionState | null> {
  return await updateSession(sessionId, {
    currentLineIndex: lineIndex,
  });
}

/**
 * Increment/decrement line index
 */
export async function nextLine(sessionId: string): Promise<SessionState | null> {
  const session = await getSession(sessionId);
  if (!session) {
    return null;
  }

  const nextIndex = Math.min(
    session.currentLineIndex + 1,
    session.currentSong?.lyrics?.length ?? 0
  );

  return await updateSession(sessionId, {
    currentLineIndex: nextIndex,
  });
}

/**
 * Decrement line index
 */
export async function prevLine(sessionId: string): Promise<SessionState | null> {
  const session = await getSession(sessionId);
  if (!session) {
    return null;
  }

  const prevIndex = Math.max(session.currentLineIndex - 1, 0);

  return await updateSession(sessionId, {
    currentLineIndex: prevIndex,
  });
}

/**
 * Update playing state
 */
export async function updateSessionPlaying(
  sessionId: string,
  isPlaying: boolean
): Promise<SessionState | null> {
  return await updateSession(sessionId, {
    isPlaying,
  });
}

/**
 * Update session settings
 */
export async function updateSessionSettings(
  sessionId: string,
  settings: Partial<DisplaySettings>
): Promise<SessionState | null> {
  const session = await getSession(sessionId);
  if (!session) {
    return null;
  }

  return await updateSession(sessionId, {
    settings: {
      ...session.settings,
      ...settings,
    },
  });
}

// ============================================================================
// Session Discovery
// ============================================================================

/**
 * Get all active session IDs
 */
export async function getActiveSessionIds(): Promise<string[]> {
  // This would require SCAN operation for production
  // For now, return empty array as we track sessions via Socket.IO rooms
  return [];
}

// ============================================================================
// Session Validation
// ============================================================================

/**
 * Check if a session exists
 */
export async function sessionExists(sessionId: string): Promise<boolean> {
  const session = await getSession(sessionId);
  return session !== null;
}

/**
 * Validate session access
 */
export async function canAccessSession(
  _sessionId: string,
  _userId: string | null
): Promise<boolean> {
  // 目前允許存取任何 session
  // 未來可能需要檢查使用者是否擁有該 session
  return true;
}

File: backend/internal/redis/session.go
// Package redis 封裝 Redis 客戶端操作。
// 此檔案負責使用 Redis 管理 session 資料，格式與 Node.js 端完全相容。
//
// Redis key 格式：
//   - session:{sessionId}          — JSON 字串，儲存 SessionState
//   - session:clients:{sessionId}  — Redis SET，每個成員為 ClientInfo 的 JSON 字串
//   - TTL: 3600 秒（1 小時）
package redis

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	goredis "github.com/redis/go-redis/v9"
)

const (
	sessionPrefix        = "session:"
	sessionClientsPrefix = "session:clients:"
	sessionTTL           = 3600 * time.Second // 1 小時
)

// SessionSong 為 session 中的歌曲資料，JSON tag 與 Node.js 格式完全一致。
type SessionSong struct {
	ID            string    `json:"id"`
	Title         string    `json:"title"`
	Artist        *string   `json:"artist"`
	Lyrics        []string  `json:"lyrics"`
	LrcTimestamps []float64 `json:"lrcTimestamps"`
	Language      *string   `json:"language"`
	UserID        string    `json:"userId"`
	CreatedAt     string    `json:"createdAt"` // ISO 8601 字串
	UpdatedAt     string    `json:"updatedAt"` // ISO 8601 字串
}

// SessionSettings 為 session 中的顯示設定，JSON tag 與 Node.js 格式完全一致。
type SessionSettings struct {
	DisplayLines    int     `json:"displayLines"`
	FontSize        int     `json:"fontSize"`
	FontFamily      string  `json:"fontFamily"`
	LineSpacing     float64 `json:"lineSpacing"`
	Theme           string  `json:"theme"`
	ShowBackground  bool   `json:"showBackground"`
	BackgroundColor string `json:"backgroundColor"`
	TextColor       string `json:"textColor"`
	HighlightColor  string `json:"highlightColor"`
	AutoScroll      bool   `json:"autoScroll"`
	ScrollDuration  int    `json:"scrollDuration"`
	EnableAnimation bool   `json:"enableAnimation"`
}

// SessionState 為完整 session 狀態，JSON tag 與 Node.js 格式完全一致。
type SessionState struct {
	SessionID        string          `json:"sessionId"`
	CurrentSong      *SessionSong    `json:"currentSong"`      // 可為 null
	CurrentLineIndex int             `json:"currentLineIndex"`
	IsPlaying        bool            `json:"isPlaying"`
	Settings         SessionSettings `json:"settings"`
	ControllerCount  int             `json:"controllerCount"`
	DisplayCount     int             `json:"displayCount"`
	CreatedAt        int64           `json:"createdAt"` // Unix 毫秒
	UpdatedAt        int64           `json:"updatedAt"` // Unix 毫秒
}

// ClientInfo 為連線客戶端資訊，JSON tag 與 Node.js 格式完全一致。
type ClientInfo struct {
	ClientID string  `json:"clientId"`
	Role     string  `json:"role"`     // "controller" | "display" | "admin"
	UserID   *string `json:"userId"`   // 可為 null
	JoinedAt int64   `json:"joinedAt"` // Unix 毫秒
}

// GetSession 從 Redis 取得指定 session。
// 若 session 不存在，回傳 (nil, nil)。
func (c *Client) GetSession(ctx context.Context, sessionID string) (*SessionState, error) {
	key := sessionPrefix + sessionID
	data, err := c.rdb.Get(ctx, key).Result()
	if err == goredis.Nil {
		return nil, nil // session 不存在
	}
	if err != nil {
		return nil, fmt.Errorf("redis get session: %w", err)
	}

	var state SessionState
	if err := json.Unmarshal([]byte(data), &state); err != nil {
		return nil, fmt.Errorf("unmarshal session: %w", err)
	}

	return &state, nil
}

// SetSession 將 session 狀態寫入 Redis，TTL 為 1 小時。
func (c *Client) SetSession(ctx context.Context, state *SessionState) error {
	key := sessionPrefix + state.SessionID
	data, err := json.Marshal(state)
	if err != nil {
		return fmt.Errorf("marshal session: %w", err)
	}

	return c.rdb.Set(ctx, key, data, sessionTTL).Err()
}

// DeleteSession 刪除 session 及其關聯的 clients 集合。
// 使用 pipeline 確保兩個 key 在同一次 round-trip 中刪除。
func (c *Client) DeleteSession(ctx context.Context, sessionID string) error {
	pipe := c.rdb.Pipeline()
	pipe.Del(ctx, sessionPrefix+sessionID)
	pipe.Del(ctx, sessionClientsPrefix+sessionID)
	_, err := pipe.Exec(ctx)
	return err
}

// GetSessionClients 取得 session 的所有已連線客戶端。
// 無法反序列化的成員會被靜默跳過。
func (c *Client) GetSessionClients(ctx context.Context, sessionID string) ([]ClientInfo, error) {
	key := sessionClientsPrefix + sessionID
	members, err := c.rdb.SMembers(ctx, key).Result()
	if err != nil {
		return nil, fmt.Errorf("redis smembers: %w", err)
	}

	clients := make([]ClientInfo, 0, len(members))
	for _, m := range members {
		var info ClientInfo
		if err := json.Unmarshal([]byte(m), &info); err != nil {
			continue // 跳過無法解析的 client
		}
		clients = append(clients, info)
	}

	return clients, nil
}

// AddClient 將客戶端加入 session 的 clients 集合。
func (c *Client) AddClient(ctx context.Context, sessionID string, info ClientInfo) error {
	key := sessionClientsPrefix + sessionID
	data, err := json.Marshal(info)
	if err != nil {
		return fmt.Errorf("marshal client info: %w", err)
	}

	return c.rdb.SAdd(ctx, key, data).Err()
}

// HasClients 檢查指定 session 是否還有已連線的客戶端
func (c *Client) HasClients(ctx context.Context, sessionID string) (bool, error) {
	key := sessionClientsPrefix + sessionID
	count, err := c.rdb.SCard(ctx, key).Result()
	if err != nil {
		return false, err
	}
	return count > 0, nil
}

// RemoveClient 依 clientID 從 session 的 clients 集合中移除對應客戶端。
// 直接使用 SMembers 原始字串做 SREM，避免重新序列化 JSON 時因 key 順序不同導致 SREM 靜默失敗。
func (c *Client) RemoveClient(ctx context.Context, sessionID string, clientID string) error {
	key := sessionClientsPrefix + sessionID
	members, err := c.rdb.SMembers(ctx, key).Result()
	if err != nil {
		return fmt.Errorf("redis smembers for removal: %w", err)
	}

	for _, raw := range members {
		var info ClientInfo
		if err := json.Unmarshal([]byte(raw), &info); err != nil {
			continue
		}
		if info.ClientID == clientID {
			// 用原始 raw 字串做 SREM，確保完全匹配
			return c.rdb.SRem(ctx, key, raw).Err()
		}
	}

	return nil
}
